packages/std/src/coins.rs

```diff
@@ -0,0 +1,246 @@
+use std::any::type_name;
+use std::collections::BTreeMap;
+use std::fmt;
+use std::str::FromStr;
+
+use crate::{Coin, StdError, StdResult, Uint128};
+
+/// A collection of coins, similar to Cosmos SDK's `sdk.Coins` struct.
+///
+/// Differently from `sdk.Coins`, which is a vector of `sdk.Coin`, here we
+/// implement Coins as a BTreeMap that maps from coin denoms to amounts.
+/// This has a number of advantages:
+///
+/// - coins are naturally sorted alphabetically by denom
+/// - duplicate denoms are automatically removed
+/// - cheaper for searching/inserting/deleting: O(log(n)) compared to O(n)
+#[derive(Clone, Default, Debug, PartialEq, Eq)]
+pub struct Coins(BTreeMap<String, Uint128>);
+
+// Casting a Vec<Coin> to Coins.
+// The Vec can be out of order, but must not contain duplicate denoms or zero amounts.
+impl TryFrom<Vec<Coin>> for Coins {
+    type Error = StdError;
+
+    fn try_from(vec: Vec<Coin>) -> StdResult<Self> {
+        let vec_len = vec.len();
+
+        let map = vec
+            .into_iter()
+            .filter(|coin| !coin.amount.is_zero())
+            .map(|coin| (coin.denom, coin.amount))
+            .collect::<BTreeMap<_, _>>();
+
+        // the map having a different length from the vec means the vec must either
+        // 1) contain duplicate denoms, or 2) contain zero amounts
+        if map.len() != vec_len {
+            return Err(StdError::parse_err(
+                type_name::<Self>(),
+                "duplicate denoms or zero amount",
+            ));
+        }
```

Comment:
Why not combine the duplicates and discard the zero amounts? It's more of a friendlier handling versus a throw.

Reply:
I'm also leaning a solution where `Vec<Coin>` -> Coins cannot fail. This can be achieved by just defining the semantics to be overriding old values. `Vec<Coin>` is typically some trusted value anyways.

Reply:
> `Vec<Coin>` is typically some trusted value anyways

Wouldn't this be an argument _for_ fail-able casting instead of against it? Since `Vec<Coin>` comes from trusted sources, it having dups or zero amounts usually indicates there are serious bugs in the source or in contract execution. In this case I think an error is the preferred behavior.

Reply:
If you put it that way, yes, it would be an argument for an assertion. Something is seriously wrong, an error case no one is prepared to handle. So it could be implemented as a panic.